### PR TITLE
Add top_blockio

### LINKED ISSUE #2989: gadgets: Implement top block-io as image based one
Port the [top block-io](https://github.com/inspektor-gadget/inspektor-gadget/tree/main/pkg/gadgets/top/block-io). Requires https://github.com/inspektor-gadget/inspektor-gadget/issues/1924 to be implemented first.

### PR DESCRIPTION
**Pull request by burak-ok**
This PR adds `top_blockio` as image based gadgets

Our builtin gadget supports kernel versions `<5.17` and `>6.5`. This is done through trying to attach the tracepoints/kprobes. Currently this can't be done with image based gadgets (#3274)

Fixes #2989

EDIT:
When having an encrpyted disk with `ig` it seems that the `--host` flag is needed. The event also shows up like the following:
```
RUNTIME.CONTAINERNAME      PID MAJOR        MINOR        COMM             BYTES              US                 IO                 RW   
                           485 259          0            dmcrypt_write/2  397312             8152               10                 write
                             0 259          0                             0                  2124               1                  read
```

### CODE REVIEW

File: gadgets/top_blockio/test/top_file_blockio.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
+// Copyright 2024 The Inspektor Gadget authors
```
Comment (blanquicet) [inline gadgets/top_blockio/test/top_file_blockio.go:1]: `gadgets/top_blockio/test/top_file_blockio.go` -> `gadgets/top_blockio/test/top_blockio_test.go`. In fact, test is not even being run.

File: gadgets/top_blockio/test/top_file_blockio.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+	containerFactory, err := containers.NewContainerFactory(utils.Runtime)
+	require.NoError(t, err, "new container factory")
+	containerName := "test-top-blockio"
+	containerImage := "docker.io/library/busybox:latest"
+
+	var ns string
+	containerOpts := []containers.ContainerOption{
+		containers.WithContainerImage(containerImage),
+		containers.WithStartAndStop(),
```
Comment (blanquicet) [inline gadgets/top_blockio/test/top_file_blockio.go:69]: Not needed as container is being stopped in `t.Cleanup`. It's only useful when passing the container to `RunTestSteps`. Please fix this also in other tests.

File: gadgets/top_blockio/test/top_file_blockio.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	if utils.CurrentTestComponent == utils.KubectlGadgetTestComponent {
+		ns = utils.GenerateTestNamespaceName(t, "test-top-blockio")
+		containerOpts = append(containerOpts, containers.WithContainerNamespace(ns))
+	}
+
+	testContainer := containerFactory.NewContainer(
+		containerName,
+		"while true; do echo -n foo > bar ; sleep 1; done",
```
Comment (blanquicet) [inline gadgets/top_blockio/test/top_file_blockio.go:79]: I don't think it'll trigger a block operation. We have to probably use `dd` as in existing tests.

**Review by blanquicet — COMMENTED** [review]
Code looks good. Please fix the test and we can go.

File: gadgets/top_blockio/README.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# top_blockio
+
+The top blockio gadget provides a periodic list of input/output block device activity. This gadget requires Linux Kernel Version 6.5+
+
+Check the full documentation on https://inspektor-gadget.io/docs/latest/gadgets/top_blockio
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply gadgets/top_blockio/README.md:5]: ref https://github.com/inspektor-gadget/inspektor-gadget/pull/3279#issuecomment-2269402083

File: gadgets/top_blockio/README.mdx [hunk 5]
```diff
@@ -0,0 +1,82 @@
+---
+title: top_blockio
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+The top blockio gadget provides a periodic list of input/output block device activity. This gadget requires Linux Kernel Version 6.5+
```
Comment (mauriciovasquezbernal) [inline gadgets/top_blockio/README.mdx:8]: top blockio -> top_blockio

File: gadgets/top_blockio/gadget.yaml [hunk 6]
```diff
@@ -0,0 +1,50 @@
+name: top blockio
+description: Periodically report input/output block device activity. This gadget requires
```
Comment (mauriciovasquezbernal) [inline gadgets/top_blockio/gadget.yaml:2]: I think "This gadget requires Kernel Version 6.5+" doesn't belong to the description.

File: gadgets/top_blockio/gadget.yaml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  Linux Kernel Version 6.5+
+homepageURL: https://inspektor-gadget.io/
+documentationURL: https://inspektor-gadget.io/docs
+sourceURL: https://github.com/inspektor-gadget/inspektor-gadget/
```
Comment (mauriciovasquezbernal) [inline gadgets/top_blockio/gadget.yaml:6]: nit: please point these to the exact gadget.

File: gadgets/top_blockio/program.bpf.c [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	int minor;
+	char comm[TASK_COMM_LEN];
+};
+
+// the value of the output summary
+struct val_t {
+	__u64 bytes;
+	__u64 us;
+	__u32 io;
+};
```
Comment (mauriciovasquezbernal) [inline gadgets/top_blockio/program.bpf.c:55]: nit: what about moving this closer to the count map definition?

**Review by mauriciovasquezbernal — COMMENTED** [review]
Some initial comments. However I'm a bit worried about this gadget. 

> When having an encrpyted disk with ig it seems that the --host flag is needed. The event also shows up like the following:

I think we need to investigate it more before continuing with this port.

**Comment by burak-ok** [comment]
I am running `docker run -it --rm polinux/stress stress -d 4`:


On a machine without encrpyted disk it looks like this:
```
RUNTIME.CONTAINERNAME                          PID MAJOR       MINOR       COMM             BYTES                 US                   IO         RW    
                                                 0 8           0                            204800                2440                 1          write 
flamboyant_chatelet                          39951 8           0           stress           41050112              175632               55         write 
flamboyant_chatelet                          39950 8           0           stress           4096                  39                   1          write 
flamboyant_chatelet                          39949 8           0           stress           1310720               5146                 1          write 
flamboyant_chatelet                          39952 8           0           stress           30236672              69208                32         write 
```

Machine with an encrpyted disk, I get no output:
```
RUNTIME.CONTAINERNAME                                                            PID MAJOR       MINOR       COMM             BYTES                US                   IO         RW             
```

Machine with an encrpyted disk, with the `--host` flag:
```
RUNTIME.CONTAINERNAME                                                            PID MAJOR       MINOR       COMM             BYTES                US                   IO         RW          
                                                                             1283100 259         0           kworker/u16:4    131072               118247               32         read        
                                                                             1370774 259         0           kworker/u16:11   8192                 31958                2          read        
                                                                                   0 259         0                            0                    3985                 1          read        
                                                                                3512 259         0           ibus-extension-  4096                 2766                 1          read        
                                                                                3512 259         0           gdbus            4096                 3506                 1          read        
                                                                                 538 259         0           dmcrypt_write/2  1412915200           37096601             10854      write       
```

File: gadgets/top_blockio/program.bpf.c [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+		valp = bpf_map_lookup_or_try_init(&counts, &info, &zero);
+
+		if (valp) {
+			/* save stats */
+			valp->us += delta_us;
+			valp->bytes += startp->data_len;
+			valp->io++;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (burak-ok) [reply gadgets/top_blockio/program.bpf.c:166]: My brain defaulted to "never use goto" way of programming higher level C++ :sweat_smile:

**Comment by burak-ok** [comment]
Thank you all for the reviews :)

**Pull request merged**